ambry-store/src/main/java/com.github.ambry.store/CompactionLog.java

```diff
@@ -0,0 +1,336 @@
+/**
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.utils.CrcInputStream;
+import com.github.ambry.utils.CrcOutputStream;
+import com.github.ambry.utils.Time;
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+
+/**
+ * Represents a record of the compaction process and helps in recovery in case of crashes.
+ */
+class CompactionLog implements Closeable {
+  private static final String COMPACTION_LOG_SUFFIX = "_compaction_log";
+  private static final short VERSION_0 = 0;
+
+  /**
+   * The {@link State} of the current compaction cycle.
+   */
+  enum State {
+    PREPARE, COPY, SWITCH, CLEANUP, DONE
+  }
+
+  private final File file;
+  private final Time time;
+  private final Long startTime;
+  private final List<CycleLog> cycleLogs;
+
+  private int currentIdx = 0;
+
+  /**
+   * Used to determine whether compaction is in progress.
+   * @param dir the directory where the compaction log is expected to exist (if any).
+   * @param name the unique name for the store under compaction.
+   * @return whether compaction is in progress for the store.
+   */
+  static boolean isCompactionInProgress(String dir, String name) {
+    return new File(dir, name + COMPACTION_LOG_SUFFIX).exists();
+  }
+
+  /**
+   * Creates a new compaction log.
+   * @param dir the directory at which the compaction log must be created.
```

Comment:
Not sure what you mean by "passing in a store and the parent directory of the store should be taken as is". Could you explain?

As for passing the partition directory and the rest being derived from it - it is possible that the path and format of that changes but storeId is something that is used for metrics and identifies the store uniquely inside a single server. Could you explain what the added pros of sending in the partition directory is (storeId is easy to pass also)?